test/tests/functional/pbs_maintenance_reservations.py

```diff
@@ -172,21 +395,120 @@ def test_maintenance_reconfirm_resv(self):
+        self.server.expect(RESV, exp_attr, id=rid2)
+
+        exp_attr = {'job_state': 'R',
+                    'exec_vnode': "(%s:ncpus=1)" % self.momB.shortname}
+        self.server.expect(JOB, exp_attr, id=jid1)
+
+        exp_attr = {'job_state': 'R',
+                    'exec_vnode': "(%s:ncpus=1)" % self.momA.shortname}
+        self.server.expect(JOB, exp_attr, id=jid2)
+
+    def test_maintenance_progressive_degrade_reservation(self):
+        """
+        Test if the reservation is partially degraded by overlapping
+        maintenance reservation at first, and then do full overlap.
+        Test if the job in fully overlapped reservation will not run.
+        Two moms (-p "servers=M1,moms=M1:M2") are needed for this test.
+        """
+        now = int(time.time())
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'managers': '%s@*' % TEST_USER})
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'scheduler_iteration': 3})
+
+        if len(self.moms) != 2:
+            cmt = "need 2 mom hosts: -p servers=<m1>,moms=<m1>:<m2>"
+            self.skip_test(reason=cmt)
+
+        self.momA = self.moms.values()[0]
+        self.momB = self.moms.values()[1]
+
+        select = 'host=%s+host=%s' % (self.momA.shortname,
+                                      self.momB.shortname)
+        a1 = {'Resource_List.select': select,
+              'reserve_start': now + 60,
+              'reserve_end': now + 7200}
+        r1 = Reservation(TEST_USER, attrs=a1)
+        rid1 = self.server.submit(r1)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+        start = self.now_plus_to_rsub_format(minutes=30)
+        end = self.now_plus_to_rsub_format(hours=1, minutes=30)
+
+        moms = [self.momA.shortname]
+        ret = self.submit_maintenance(TEST_USER, start, end, moms)
+        self.assertEquals(ret['rc'], 0)
+
+        exp_attr = {'resv_nodes': '(%s:ncpus=1)' % self.momB.shortname,
+                    'reserve_state': (MATCH_RE, 'RESV_DEGRADED|12'),
+                    'reserve_substate': 12}
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+        self.logger.info("Wait for reservation to start (2 minutes)")
+        time.sleep(120)
+
+        start = self.now_plus_to_rsub_format(minutes=30)
+        end = self.now_plus_to_rsub_format(hours=1, minutes=30)
+
+        moms = [self.momB.shortname]
+        ret = self.submit_maintenance(TEST_USER, start, end, moms)
+        self.assertEquals(ret['rc'], 0)
+
+        a = {'Resource_List.select': '1:ncpus=1',
+             'Resource_List.walltime': '1000',
+             'queue': rid1.split('.')[0]}
+        j = Job(TEST_USER, attrs=a)
+        j.set_sleep_time(990)
+        jid1 = self.server.submit(j)
+
+        self.logger.info("Wait for the job to try to run (30 seconds)")
+        time.sleep(30)
```

Comment:
Why are we waiting 30s for the job to run?  If the reservation is running, a cycle should be run when the job is submitted.

Reply:
The job is expected to stay in Q and do not run because its reservation was fully overlapped. There could be a false positive result without any timeout. Maybe 10 seconds is enough, but I think 30 is bearable.